content/blog/2022/headed-playwright-in-wsl-docker.md

```diff
@@ -0,0 +1,114 @@
+# Update the image tag to match your desired version of Playwright
+docker run --rm mcr.microsoft.com/playwright:v1.28.0 npx -y playwright open google.com
+
+╔════════════════════════════════════════════════════════════════════════════════════════════════╗
+║ Looks like you launched a headed browser without having a XServer running.                     ║
+║ Set either 'headless: true' or use 'xvfb-run <your-playwright-app>' before running Playwright. ║
+║                                                                                                ║
+║ <3 Playwright Team                                                                             ║
+╚════════════════════════════════════════════════════════════════════════════════════════════════╝
+```
+
+No browser window is launched! Instead we get an error message about not "having
+a XServer running". The definition and functionality of XServer are beyond the
+scope of this article, but let's say that without it we can't interact with
+applications that require a user interface (like the browser). Feel free to dive
+into [a more detailed explanation](https://askubuntu.com/a/7885/27669) if you
+prefer.
+
+Searching for this error on the web will return results explaining how to
+install and start XServer. That advice applies to non-containerized, Linux-based
+systems. Our host system is Windows, so we actually don't want to do that.
+Instead we want the container to use the *host XServer* to launch Playwright
+*inside the container*.
+
+You might find it surprising (I certainly did) to find out that Microsoft
+Windows has a native XServer even though its not a GNU/Linux system. It's called
+[WSLg](https://github.com/microsoft/wslg#readme) and it's included as part of
+the [Windows Subsystem for Linux](https://aka.ms/wslstorepage) (WSL). You most
+likely already have WSL and WSLg installed if you are running [Docker
+Desktop](https://www.docker.com/products/docker-desktop/) in recent builds of
+Windows 10 and 11.
+
+Let's start by verifying that WSL and WSLg are installed and running:
+
+```powershell
+wsl.exe --version
+
+WSL version: 1.0.0.0
+Kernel version: 5.15.74.2
+WSLg version: 1.0.47
+MSRDC version: 1.2.3575
+Direct3D version: 1.606.4
+DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
+Windows version: 10.0.22621.819
+```
+
+If the command returns without errors and you see a line with `WSLg version:
+X.X.X`, then you're good to go! According to [this
+guide](https://github.com/microsoft/wslg/blob/main/samples/container/Containers.md),
+we only need to do two things:
+
+1. Set the `DISPLAY` environment variable inside the container.
+2. Mount the XServer Unix socket inside the container.
+
+Docker allows us to do both with a couple command line flags:
+
+```bash
+docker run --rm \
+-e DISPLAY=:0 \ # Set the env var
+-v /tmp/.X11-unix:/tmp/.X11-unix \ # Mount the socket
+mcr.microsoft.com/playwright:v1.28.0 npx -y playwright open google.com
```

Comment:
@dvdherron could you verify this actually works on your Windows machine? 😬

Reply:
Looks like the terminal interpreted that as four different commands, not one command split into four lines. Did you include the trailing backslashes?

Reply:
Oh, it's the comments, they are breaking the newline escaping. I'll rewrite that

Reply:
Fixed in 4fc8e4634c79728f5160cb9f4b274b8b0e383ee1